File: medium/215.ts
/**
    215. Kth Largest Element in an Array

    方法:   
        counting sort
        輸入有負數，所以要調成從0開始(減掉min)
 */

function findKthLargest(nums: number[], k: number): number {
    const min = nums.reduce((min, v) => Math.min(min, v), 100001)
    const max = nums.reduce((max, v) => Math.max(max, v), 0)
    const count: number[] = new Array(max - min + 1).fill(0)
    nums.forEach(v => count[v - min]++)

    let i = count.length - 1
    while (i >= 0) {
        k -= count[i]
        if (k <= 0) return i + min
        --i
    }

    return -1
};

File: medium/215.java
/**
    215. Kth Largest Element in an Array

    counting sort approach

    Time complexity: O(n + k)
    Space complexity: O(k)
    n = |nums|, k = 200001
 */
class Solution {
    public int findKthLargest(int[] nums, int k) {
        int offset = 10000;
        int N = offset * 2 + 1;
        int[] cnt = new int[N];

        for(int i = 0; i < N; ++i)
            cnt[i] = 0;
        
        for(int num : nums)
            cnt[num + offset]++;
        
        int j = 0;
        for(int i = N - 1; i >= 0; --i){
            j += cnt[i];
            if(j >= k) return i - offset;
        }

        return -1;
    }
}
